pkg/flowaggregator/certificate.go

```diff
@@ -0,0 +1,180 @@
+			DNSNames:    []string{flowAggregatorDNSName},
+			NotBefore:   validFrom,
+			NotAfter:    validFrom.Add(maxAge),
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+			KeyUsage:    x509.KeyUsageDigitalSignature,
+		}
+	} else {
+		cert = &x509.Certificate{
+			SerialNumber: big.NewInt(3),
+			Subject: pkix.Name{
+				CommonName: fmt.Sprintf("flow-aggregator-client-certificate@%d", time.Now().Unix()),
+			},
+			NotBefore:   validFrom,
+			NotAfter:    validFrom.Add(maxAge),
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+			KeyUsage:    x509.KeyUsageDigitalSignature,
+		}
+	}
+	// generate private key for certificate
+	certKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, err
+	}
+	// sign the certificate using CA certificate and key
+	certBytes, err := x509.CreateCertificate(rand.Reader, cert, caCert, &certKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPEM := new(bytes.Buffer)
+	pem.Encode(certPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+
+	certKeyPEM := new(bytes.Buffer)
+	pem.Encode(certKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(certKey),
+	})
+
+	return certPEM.Bytes(), certKeyPEM.Bytes(), nil
+}
+
+func syncCAAndClientCert(caCert, clientCert, clientKey []byte, k8sClient kubernetes.Interface) error {
+	klog.Info("Syncing CA certificate with ConfigMap")
+	caConfigMap, err := k8sClient.CoreV1().ConfigMaps(CAConfigMapNamespace).Get(context.TODO(), CAConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting ConfigMap %s: %v", CAConfigMapName, err)
+	}
+	if caConfigMap.Data != nil && caConfigMap.Data[CAConfigMapKey] == string(caCert) {
+		return nil
+	}
+	caConfigMap.Data = map[string]string{
+		CAConfigMapKey: string(caCert),
+	}
+	if _, err := k8sClient.CoreV1().ConfigMaps(CAConfigMapNamespace).Update(context.TODO(), caConfigMap, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("error updating ConfigMap %s: %v", CAConfigMapName, err)
+	}
+
+	secret, err := k8sClient.CoreV1().Secrets(ClientSecretNamespace).Get(context.TODO(), ClientSecretName, metav1.GetOptions{})
+	exists := true
+	if err != nil {
+		exists = false
+		secret = &v1.Secret{
+			Data: map[string][]byte{
+				"tls.crt": clientCert,
+				"tls.key": clientKey,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ClientSecretName,
+				Namespace: ClientSecretNamespace,
+			},
+			Type: v1.SecretTypeTLS,
+		}
+	}
+	secret.Data = map[string][]byte{
+		"tls.crt": clientCert,
+		"tls.key": clientKey,
+	}
+	if exists {
+		if _, err := k8sClient.CoreV1().Secrets(ClientSecretNamespace).Update(context.TODO(), secret, metav1.UpdateOptions{}); err != nil {
+			return fmt.Errorf("failed to update Secret %s: %v", ClientSecretName, err)
+		}
+	} else {
+		if _, err := k8sClient.CoreV1().Secrets(ClientSecretNamespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
```

Comment:
I'm little confused, so it assumes the secret may not be present. But it's created in the yaml, why we do this differently from configmap?
But maybe you need to keep the create logic because of the field owner issue above.

Reply:
yes it should be removed if we create in the yaml. I removed the creation code in the yaml file. Thanks